src/openads-appnexus/repository/AdListenerRepository.js

```diff
@@ -0,0 +1,51 @@
+import ReplayEventBus from '../service/ReplayEventBus'
+
+const AD_STORED = 'AD_STORED'
+
+const DEFAULT_TIMEOUT = 20000
+const TIMEOUT_EXCEPTION = 'Timeout waiting to retrieve the Ad'
+
+export default class AdListenerRepository {
+  constructor({timeout = DEFAULT_TIMEOUT} = {}) {
+    this._timeout = timeout
+  }
+
+  find({id}) {
+    return Promise.race([
+      new Promise((resolve, reject) => {
+        const eventId = this._eventId({id})
+        ReplayEventBus.register({
+          eventName: eventId,
+          observer: ({event, payload}) => resolve(payload)
```

Comment:
When resolve is called, we should remove the observer right?
Second time you call find method will not work because "resolve" will have a reference to an old promise already solved.
In Summary as we already done in CMP, find is like a transaction block, you add the listener and when is called, you remove it ;)